proxy/src/main/java/net/md_5/bungee/ServerConnection.java

```diff
@@ -25,9 +25,11 @@
     private boolean isObsolete;
     @Getter
     private final boolean forgeServer = false;
+    /**
+     * Add sent keepalives at the end, so the oldest keep alive expected to be recieved next is at the start
+     */
     @Getter
-    @Setter
-    private long sentPingId = -1;
+    private final LinkedList<KeepAliveInfo> sentKeepAlives = new LinkedList<>();
```

Comment:
Queue needs to be size limited otherwise its easily exploitable

Reply:
Oh right

Reply:
Put in a time-based check in DownstreamBridge, that prevents clients from not answering keep alives in time.